hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowOrdinalMapper.java

```diff
@@ -0,0 +1,196 @@
+
+import static com.netflix.hollow.core.HollowConstants.ORDINAL_NONE;
+import java.util.Arrays;
+import java.util.HashMap;
+
+public class HollowOrdinalMapper {
+    private int size = 0;
+    private static final double LOAD_FACTOR = 0.7;
+    
+    private Integer[] ordinalMappings;
+    private Integer[] originalHash;
+
+    private HashMap<Integer, Object[]> indexFieldObjectMapping;
+    private final HashMap<Integer, Integer> assignedOrdinalToIndex;
+
+    private final PrimaryKey primaryKey;
+    private final int[][] keyFieldIndices;
+    private final boolean[] keyFieldIsIndexed;
+
+    private final ObjectInternPool memoizedPool;
+
+    public HollowOrdinalMapper(PrimaryKey primaryKey, boolean[] keyFieldIsIndexed, int[][] keyFieldIndices) {
+        // Start with prime number to assist OA
+        this.ordinalMappings = new Integer[2069];
+        this.originalHash = new Integer[2069];
+        Arrays.fill(this.ordinalMappings, ORDINAL_NONE);
+
+        this.primaryKey = primaryKey;
+        this.keyFieldIndices = keyFieldIndices;
+        this.keyFieldIsIndexed = keyFieldIsIndexed;
+
+        this.indexFieldObjectMapping = new HashMap<>();
+        this.assignedOrdinalToIndex = new HashMap<>();
+
+        this.memoizedPool = new ObjectInternPool();
+    }
+
+    public int findAssignedOrdinal(HollowObjectTypeReadState typeState, int keyOrdinal) {
+        int hashedRecord = hashKeyRecord(typeState, keyOrdinal);
+        int index = indexFromHash(hashedRecord, ordinalMappings.length);
+
+        while (ordinalMappings[index]!=ORDINAL_NONE) {
+            if(recordsAreEqual(typeState, keyOrdinal, index))
+                return ordinalMappings[index];
+            index = (index + 1) % ordinalMappings.length;
+        }
+
+        return ORDINAL_NONE;
+    }
+
+    private boolean recordsAreEqual(HollowObjectTypeReadState typeState, int keyOrdinal, int index) {
+        for(int fieldIdx=0;fieldIdx<primaryKey.numFields();fieldIdx++) {
+            if(!keyFieldIsIndexed[fieldIdx])
+                continue;
+
+            Object newFieldValue = readValueInState(typeState, keyOrdinal, fieldIdx);
+            Object existingFieldValue = indexFieldObjectMapping.get(index)[fieldIdx];
+            if(!newFieldValue.equals(existingFieldValue)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public int storeNewRecord(HollowObjectTypeReadState typeState, int ordinal, int assignedOrdinal) {
+        int hashedRecord = hashKeyRecord(typeState, ordinal);
+
+        if ((double) size / ordinalMappings.length > LOAD_FACTOR) {
+            expandAndRehashTable();
+        }
+
+        int index = indexFromHash(hashedRecord, ordinalMappings.length);
+
+        // Linear probing
+        while (ordinalMappings[index] != ORDINAL_NONE) {
+            if(recordsAreEqual(typeState, ordinal, index)) {
+                this.assignedOrdinalToIndex.put(assignedOrdinal, index);
```

Comment:
Multiple `assignedOrdinal`s can map to the same index. But when `expandAndRehashTable` calls `rehashExistingRecord` it only updates one of these assignedOrdinals to point to map to the new index but some entries in `assignedOrdinalToIndex` will still be pointing to the non-remapped index value. Am I reading this right?

Reply:
With deltas, there should never be repeated `assignedOrdinals`, so in practice, this will not happen.

However, I realized that we don't empty the `ordinalMappings` table when there's a snapshot. This won't cause any errors as we still overwrite the key in `assignedOrdinalToIndex`, but it will clog up the hash table because we aren't removing values that will never be retrieved again. Should be a one-line fix that I'll implement later, good catch!

Reply:
Fixed in 1e522166